[Social Work] - Enhance technical documentation (#1705)
csg-org/CompactConnect: backend/social-work-app/docs/design/README.md

single-state license in the same jurisdiction and license type (see
[Privilege Runtime Generation for Multi-State Licenses](#privilege-runtime-generation-for-multi-state-licenses)), a
single-state-only licensee will never have privileges generated for them. Their data is still valuable to the
system and to other member states, since it establishes the practitioner's licensure record and status within the
compact even though they are not (yet) authorized to practice in other jurisdictions.

If a jurisdiction uploads a multi-state license as compact-eligible while its paired single-state license in the same jurisdiction is already ineligible (inactive, expired, or encumbered), the multi-state license is **still persisted** rather than rejected, and a
`license.validation-error` event is published so the uploading jurisdiction can be notified of the mismatch (see
`_check_for_multi_state_single_state_eligibility_validation_error` in
[ingest.py](../../lambdas/python/provider-data-v1/handlers/ingest.py)). This ensures that licensee data is always
accepted and ingested even in cases where eligibility for compact privileges cannot yet be fully confirmed.

> Re-uploading that same license (determined by providerId, jurisdiction, licenseType, and licenseScope) updates the existing record in place (rather than creating a duplicate), and, when relevant fields changed, creates a `licenseUpdate` history record categorized as a `renewal` (when the expiration date moved forward — i.e., the jurisdiction reissued the license for a new term), a `deactivation` (when the jurisdiction reports the license as inactive), or `other` (for any other changed field).

Author: I have added details about these update categories to the respective encumbrance and investigation sections where the information is more relevant.

This architecture ensures that SSN data is protected throughout the ingest process while still allowing the system to
associate licenses with the correct providers across jurisdictions.

### All-Licensee Ingestion, Including Single-State Licensees

The ingest pipeline described above processes **every uploaded licensee record identically**, regardless of the
license's `licenseScope` (`single-state` or `multi-state`, see [License Scopes](#license-scopes)). A jurisdiction is

> not required to upload only its multi-state licensees; it is expected to upload data for **all** of its eligible

Reviewer: This sort of implies jurisdictions do not _have_ to upload its single-state licenses, but I think you are saying they must?  This might be more clearly stated in 'must', 'should', and 'may' terms.

> A jurisdiction must upload data for all of its eligible providers, including all single-state and multi-state licenses associated with each provider.

Maybe? I changed to use the term 'providers' because reading it quickly led me to read it as licenses, rather than licensees.

Author: I have reworded this to make it more clear that we expect states to upload both license types for all of the providers they intend to upload information for.

### All-Licensee Ingestion, Including Single-State Licensees

The ingest pipeline described above processes **every uploaded licensee record identically**, regardless of the
license's `licenseScope` (`single-state` or `multi-state`, see [License Scopes](#license-scopes)). It is expected that jurisdictions will upload all single-state and multi-state licenses associated with each provider that they intend to upload data for.

> Single-state-only licensees are fully ingested and stored as `license` records. However, because privilege generation requires a **paired** multi-state and

Reviewer: licenses here, not licensees, right?

Reviewer: Oh wait, did you mean something like "Single-state licenses are fully ingested and stored as `license` records, with or without a paired multi-state license."?

Author: I've reworded this to hopefully make it a little more clear.

Generation](#multi-state-license-model--privilege-generation)). Per Commission rule, the CUID will be:
- **Issued only when the practitioner's home state authorizes a Multistate License** — i.e. generated the first time
  a practitioner has an eligible multi-state license on file, rather than at first license ingestion in general (contrast with the RID, which is assigned on first license ingestion regardless of scope).
- **Public-facing**, unlike the RID, and required by Commission rule to be shown in public lookup.
- **Person-based, not license-based**, and persistent for mobility: the CUID will remain the same even if the practitioner's home state later changes under Compact rules (see [Home State Changes](#home-state-changes)).
- **Permanently linked to the practitioner's RID** (`providerId`), so historical and future records remain consistent regardless of any home-state changes.

Once implemented, the CUID will be stored as the optional `publicCompactIdentifier` field on the top-level
[Provider Record](#record-types-in-detail). Generation is planned as part of the license ingest flow in

> [ingest.py](../../lambdas/python/provider-data-v1/handlers/ingest.py): whenever a license a multi-state license is uploaded for a practitioner, the ingest handler will check whether `publicCompactIdentifier` is already set on the practitioner's provider record. If it is not yet set, a new CUID will be generated at that point and written to the provider record. Because the check is against the existing value on the provider record, the CUID is generated exactly once and is never regenerated on subsequent multi-state license uploads, including after a [home state change](#home-state-changes). Because it will live on the top-level provider record rather than on an individual license, the CUID will also be added as a top-level field in the [OpenSearch index mapping](#index-mapping), making practitioners searchable by CUID (see [Advanced Data Search](#advanced-data-search)).

Author: I just clarified with the client that the CUID will be generated whenever any multi-state license is uploaded in to the system for a practitioner.

   - `licenseScope` identifies the target license when `investigationAgainst` is `license`
   - Privilege-targeted investigations always use `licenseScope` `single-state`, since privileges are scoped to a particular state.

> ### Commission Unique Identifiers: RID and CUID

Reviewer: Despite being part of some of these discussions, I'm still a bit confused by what the CUID provides that the RID does not, except for it only being generated when the provider is eligible to practice in compact states. But why could the RID just be 'revealed' publicly at the same point when the CUID would otherwise be generated?

Author: That was my argument as well, but apparently the commission is insistent that they are legally required to have a public id distinct from an internal id. @KbisonCSG has informed us that the commission will almost certainly be implementing this feature despite alternative options to only make the practitioner visible in the public search if they are able to practice in multiple states, so this is our proposed solution for supporting a CUID.

state submissions to the system, and maps directly to the system's `providerId` — the identifier generated the
first time a license is ingested for a practitioner (see [License Preprocessing](#ingest-flow)). The RID
(`providerId`) is:
- **Immutable and internal only**: once assigned it never changes, and it is not displayed in the public lookup
  portal.
- **Visible to staff users**: although hidden from the public portal, the RID is visible to staff users through the staff user search.

#### Compact Unique Identifier (CUID) — Planned

The CUID is the Compact's persistent, public-facing identifier for individuals who hold multistate
authorization (i.e. at least one generated privilege — see [Multi-State License Model / Privilege
Generation](#multi-state-license-model--privilege-generation)). Per Commission rule, the CUID will be:
- **Issued only when a state uploads a Multistate License** — i.e. generated the first time a practitioner has a multi-state license uploaded into the system, rather than at first license ingestion in general (contrast with the RID, which is assigned on first license ingestion regardless of scope).

> - **Public-facing**, unlike the RID, and required by Commission rule to be shown in public lookup.

Reviewer: ```suggestion
- **Public-facing**, unlike the RID, required by Commission rule to be shown in public lookup.
```

> - **Person-based, not license-based**, and persistent for mobility: the CUID will remain the same even if the practitioner's home state later changes under Compact rules (see [Home State Changes](#home-state-changes)).

Reviewer: Isn't this true of the RID, too?

How scopes are represented in DynamoDB sort keys and related record types is documented under
[License Composite Identifier](#license-composite-identifier) in the Data Model section.

> ### MSL Issuance and Home-State Confirmation

Reviewer: Can we spell this out for the header: "Multi-state License"

In practice these usually agree (most practitioners hold only one license type), but they are calculated separately
and in theory can diverge for practitioners holding multiple license types with different upload/renewal histories.

**Eligibility confirmation:** When a jurisdiction uploads a license, its reported status and eligibility are
persisted as-is (`jurisdictionUploadedLicenseStatus`, `jurisdictionUploadedCompactEligibility`). The system does not
treat those jurisdiction-reported values as final; on every load, `licenseStatus` and `compactEligibility` are
**recalculated** from the jurisdiction-reported values, the current date versus the expiration date, and the
license's encumbrance state (see [Status Calculation](#security-and-status-calculation) in the Data Model section).
An MSL is only eligible for privilege generation once its calculated `compactEligibility` is `eligible`.

Once a home MSL is confirmed eligible for a given license type, remote-state Multistate Authorization to Practice (privileges) are generated for that license type as described next.

### Privilege Runtime Generation for Multi-State Licenses

> Privileges are generated from a multi-state home license: one privilege per compact member jurisdiction (other than the home jurisdiction) for that license type. If the most recent multi-state license is ineligible or its associated single-state license is ineligible, no privileges are generated for that license type (there is no fallback to an older jurisdiction). This means that privileges are only returned from the API for a practitioner when the most recently issued or renewed multi-state license is the one with the pairing.

Reviewer: Couple small suggested edits for clarity.

```suggestion
Privileges are generated from a multi-state home license: one privilege per compact member jurisdiction (other than the home jurisdiction) for that license type. If the most recent multi-state license is ineligible or its associated single-state license is ineligible, no privileges are generated for that license type (there is no fallback to an older multi-state license from another jurisdiction). This means that privileges are only returned from the API for a practitioner when the most recently issued or renewed multi-state license is eligible and paired with an eligible single-state license.
```

that privilege jurisdiction and license type. License-targeted adverse actions and investigations additionally specify
which `licenseScope` they apply to.

### Home State Changes

> A home jurisdiction change occurs when, after a license is ingested, the most recent multi-state license with a paired single-state license is from a **different** jurisdiction than the practitioner's current home state on the provider record. When this is detected:

Reviewer: ```suggestion
A provider's home jurisdiction change occurs when, after a license is ingested, the most recent multi-state license (across all license types) with a paired single-state license is from a **different** jurisdiction than the practitioner's current home state on the provider record. When this is detected:
```

1. A `provider.homeStateChange` event is published to the data event bus (see [Notifications](#notifications)), which results in an email notification to the **former** home jurisdiction only.
2. The practitioner's top-level provider record is rebuilt using the data from the new home license (name, contact information, `licenseJurisdiction`, etc.).
3. A `providerUpdate` history record is written with `updateType: homeJurisdictionChange`, preserving the previous provider-record values.
4. Because privileges are not stored and are instead generated fresh on every read (see
   [Privilege Runtime Generation](#privilege-runtime-generation-for-multi-state-licenses)), privileges tied to the
   **old** home jurisdiction stop being generated as soon as the change takes effect — there is no persisted privilege

>    record to separately deactivate, and no grace period.

Reviewer: Hmm, is this 100% true? Your earlier note mentions the (rare) case where a provider can have multiple licenses with multiple home states, such that the privilege-generation home-state for a particular license type is different from the provider's saved home state.

Maybe we mean to say here, "...privileges tied to the **old** home jurisdiction for the license type that triggered the provider home jurisdiction change stop being generated as soon..."

(That is wordy, there's probably a better phrasing...)

Author: I have reworded this to hopefully clarify what the actual behavior is.

**One-home-state-per-license-type enforcement:** For a given license type, the home MSL is always the single most
recently issued/renewed **paired** multi-state license (sorted by `dateOfRenewal`, falling back to `dateOfIssuance`,

> with `dateOfIssuance` as a final tiebreaker). Notably, this selection does **not** consider whether the candidate

Reviewer: > (sorted by `dateOfRenewal`, falling back to `dateOfIssuance`, with `dateOfIssuance` as a final tiebreaker)

Just confirming this is right– `dateOfIssuance` listed twice?

Author: It is awkwardly worded. The intent was to call out that state that if the `dateOfRenewal` is the same for both licenses, the `dateOfIssuance` is the fallback and tiebreaker. I'll find a better way to reword this.

2. **Privilege Jurisdictions**: Other compact member states where the provider can practice under the compact; one generated privilege per (jurisdiction, license type).
3. **License-Based Eligibility**: Privileges are derived from stored licenses and are only generated when the chosen home license for that type is valid and compact-eligible. Status is then modified by adverse actions and investigations.

### Deactivation Effects from Adverse Action Processing

The system supports encumbering (taking adverse action against) both licenses and privileges. Encumbrance processing is split across two layers:

1. **Synchronous data writes**: The staff admin API (`encumbrance.py`) validates the request and, through the data
   access layer, writes an `adverseAction` record and updates the relevant `encumberedStatus` field(s) in DynamoDB.
2. **Asynchronous notifications**: Once the write succeeds, an event is published to the data event bus. The
   corresponding data-events handler (`encumbrance_events.py`) only sends email notifications to affected states
   (see [Notifications](#notifications)) — it does not perform any further data mutation.

> `encumberedStatus` is the only value actually persisted by an encumbrance action. As described in

Reviewer: Do we mean to say... 
```suggestion
`encumberedStatus` is the only value actually persisted by an encumbrance action in the target license record. As described in
```

Or is it the provider record? Since privileges don't exist in the DB, where does `encumberedStatus` live for those?

Author: encumberedStatus is specific to license records in this case, since there is a persistent record in the DB. The encumberedStatus of privileges are determined by the adverse action records associated for that particular state/license type, which are persisted in the DB. I will add this detail to the document.

[Status Calculation](#security-and-status-calculation), `licenseStatus` and `compactEligibility` are always
**derived** from `encumberedStatus` (along with expiration and jurisdiction-reported status) at load time, rather
than being written directly. Encumbrance and lift actions each write a categorized `licenseUpdate` history record
with `updateType: encumbrance` or `lifting_encumbrance`, respectively.

The following four scenarios describe how encumbering a license or privilege affects deactivation, depending on what
is targeted:

| Scenario | Effect on generated privileges |
|---|---|

> | Home-state **single-state** license encumbered | The paired home multi-state license is displayed as ineligible (its eligibility is inherited from the single-state license, see [Privilege Runtime Generation](#privilege-runtime-generation-for-multi-state-licenses)), so **all** privileges generated for that license type are suppressed |

Reviewer: To clarify, if the single-state license is encumbered, _both_ the multi-state and single-state licenses are marked as 'ineligible'. Why not just mark the single-state license as ineligible and rely on the privilege generation logic that requires both to be eligible?

Author: That is how we do it. We do not mark the multi-state license as ineligible in the DB, only display it as ineligible: 

> The paired home multi-state license is **displayed** as ineligible

Reviewer: Ah, got it. Thanks!

| Home-state **multi-state** license encumbered | The home multi-state license itself becomes ineligible. Because privilege generation requires **both** the home multi-state license and its paired single-state license to be eligible, the still-eligible single-state license cannot rescue privilege generation — **all** privileges for that license type are suppressed, the same as above |

> | **Remote** (non-home) state license encumbered | **No effect** on generated privileges. This is true regardless of whether the remote license's `licenseScope` is `single-state` or `multi-state` — both scopes are encumbered through the identical code path, so behavior is the same. The encumbrance is scoped to that one license record (and it flags the provider as encumbered compact-wide); it only starts to matter for privilege generation if that license later becomes the practitioner's home license |

Reviewer: To clarify, if I have a MSL-SSL pairing in two states (say, OH and KY), OH is my home state (more recent MSL), and my KY license is encumbered, I can still get privileges to practice in the compact states?

Would I still get a privilege to practice in KY? Would that not just override the encumbrance that was placed on my KY license?

Author: So with the way the system is coded today, yes the system would display an active privilege for KY as a result of the eligible OH license, even if the KY license of the same type is encumbered.

This is a great question which we will need to clarify with @KbisonCSG. @KbisonCSG - could you verify what the behavior should be in this case?

Reviewer 2: @landonshumway-ia & @jlkravitz Technically, yes since their eligibility is only based on encumberance of the home state license (OH) in this example. However, I would say this would be a very niche case as i dont see a reason for someone to keep an SSL in a state they are elgible to work in through the compact anyways, unless it is just leftover for now while they are moving between home states. In that case, I also think it would be very rare for another state to issue a license to someone with an encumbered licese in another state.

If for some reason a SW additionally had an SSL in KY and it was encumbered, my thinking is that licensing board would then decide to revoke their privlege, which they could then go in and do on a case-by-case basis, but it is possible that they might want the data system to do so automatically? If possible, I can ask the admins tommorrow if that is something they would like to have happen.

Reviewer: You and Landon are the SMEs here, so I'll defer to your read on complexity/feasibility/frequency/etc. In my experience, though, rare cases like this tend to surface more often than expected as systems scale, so I'd lean toward scoping rules that handle them explicitly rather than leaving them to admins to handle on a case-by-case basis.

Author: @jlkravitz - to @KbisonCSG's point, in this particular edge case the state admin for KY could encumber the individual privilege record in KY if needed, so we do have a path forward if it does occur.  

That being said @KbisonCSG, it would be good to clarify this behavior with the admins if possible so we can automate that deactivation now if needed before the system is live.

Author: @jlkravitz I have received feedback on this from @KbisonCSG and added the details in this commit https://github.com/csg-org/CompactConnect/pull/1705/changes/6b3873e0334ee379ded3c461802ecc4ac9aedbd7

Returns one result row per indexed document (one per license). Each hit is a full provider-shaped document for that license row (including the single license in `licenses` and generated privileges as applicable).

#### Public Provider Search
```
POST /v1/public/compacts/{compact}/providers/query
```

An unauthenticated endpoint (implemented in
[public_search.py](../../lambdas/python/search/handlers/public_search.py)) that allows members of the public to

> search for practitioners by license number, jurisdiction, family name, and/or given name. Like Provider Search, it

Reviewer: Do you mean "Like the non-public Provider Search endpoint..."?

Author: Yes, I've reworded that section for clarity.

returns one result row per matching indexed document (one per license), sanitized down to only the fields intended
for public display.

The public query filters documents down to a practitioner's home multi-state license per license type
(`licenses.licenseScope: multi-state` and `licenses.mostRecentLicenseForType: true`), consistent with the fact that
only a home multi-state license can currently generate privileges.

**Planned:** once the CUID is implemented (see [Commission Unique Identifiers: RID and
CUID](#commission-unique-identifiers-rid-and-cuid)), the public query will add an additional filter requiring the
top-level `publicCompactIdentifier` field to be populated (i.e. non-null/non-empty) on the document. Since the CUID
is only generated once a practitioner has a paired single-state/multi-state license on file (see [Compact Unique
Identifier (CUID)](#compact-unique-identifier-cuid--planned)), this filter has the effect of excluding practitioners

> who have not yet met that pairing requirement from public search results entirely. This ensures the public can only find and reference practitioners using a

Reviewer: To confirm my understanding: All practitioners will live in OpenSearch, but the public API will _require_ a CUID provided, such that the public search is really only usable for searching for a specific provider's multi-state licenses?

Author: Correct, the public search will only match against practitioners that have both a single and multi-state license pairing. Until both of those records are uploaded, we will not generate a CUID and the practitioner will not be visible in the public search. I have reworded this section slightly to hopefully make this clearer.
